attachment_queue/models/attachment_queue.py

```diff
@@ -115,18 +117,38 @@ def run_as_job(self):
                 seconds=DEFAULT_ETA_FOR_RETRY,
                 ignore_retry=True,
             ) from exc
-        if self.state == "pending":
+        if self.state != "done":
             try:
                 with self.env.cr.savepoint():
                     self.run()
-            except Exception as e:
-                _logger.warning(STR_ERROR_DURING_PROCESSING.format(self.id) + str(e))
-                self.write({"state": "failed", "state_message": str(e)})
-                emails = self.failure_emails
-                if emails:
-                    self.env.ref(
-                        "attachment_queue.attachment_failure_notification"
-                    ).send_mail(self.id)
+            except OperationalError as err:
+                # re-raise typical transaction serialization error so queue job retries
+                # no need to set attachment as failed since it will be retried.
+                if err.pgcode in PG_CONCURRENCY_ERRORS_TO_RETRY:
+                    raise
+                self._set_attachment_failure(err)
+            except RetryableJobError as err:
+                # re-raise Retryable Error to keep the functionality
+                # we still set the attachment as failed here because it may not be
+                # retried (in case max_retries is reached). We would not want a
+                # pending attachment with no related pending job
+                self._set_attachment_failure(err)
+                # a rollback has been made before because of the savepoint.
+                # we need to commit because we re-raise the exception and a rollback
+                # will be performed
+                self.env.cr.commit()  # pylint: disable=E8102
+                raise
+            except Exception as err:
+                self._set_attachment_failure(err)
+
+    def _set_attachment_failure(self, error):
+        _logger.warning(STR_ERROR_DURING_PROCESSING.format(self.id) + str(error))
```

Comment:
I could overload this but what do you think of adding at the end of this function a hook for failure side effects ?

For example your attachment.queue is supposed to do something on a sale order. Then you could use this hook in that way:
`
class SaleProcessorException(Exception):
    def __init__(self, message, record=False):
        super().__init__(message)
        self.record = record
`

on attachment.queue

`
def _run_sale_order_processing(self)
    try:
        ...
    except ValueError as e:
        raise SaleProcessorException("Error during processing or sale order xyz", sale_order)

def _error_hook(self, exc):
    e.record.attachment_queue_id = self
    e.record.attachment_queue_warning = f"An automated processing failed: str(e)"
`

Reply:
Well, I guess we could add some kind of `_error_hook`, but I am not sure why it is needed ?
`_set_attachment_failure` is already a usable hook for this no?

With your example, you just would have : 
```
def _set_attachment_failure(self, exc):
    super()._set_attachment_failure()
    if type(e) is SaleProcessorException:
        e.record.attachment_queue_id = self
        e.record.attachment_queue_warning = f"An automated processing failed: str(e)"
```

I don't see any gain of having one more hook, but I may miss something.

Reply:
It's not necessary and is a personal preference, trading some conciseness for explicitness